pkgs/dart_mcp/lib/src/api/tools.dart

```diff
@@ -290,7 +362,315 @@ extension type Schema.fromMap(Map<String, Object?> _value) {
   List<Schema>? get not => (_value['not'] as List?)?.cast<Schema>();
 }
 
+extension SchemaValidation on Schema {
+  /// Validates the given [data] against this schema.
+  ///
+  /// Returns a list of [ValidationError] if validation fails,
+  /// or an empty list if validation succeeds.
+  List<ValidationError> validate(Object? data) {
+    final failures = <ValidationError>[];
```

Comment:
How about just making this be a custom HashSet?

```
    final failures = HashSet<ValidationError>(
      equals: (a, b) => ...,
      hashCode: (a) => ...,
    );
```

Reply:
That's a good idea: would elimnate the duplicate checks.